### PR TITLE
Added granular logging for the existing provider upgrade test

### PR DESCRIPTION
**Pull request by parul5sahoo**
Added logging statements for different checks in the provider upgrade test to fetch details of failure.

**NOTE**
verified the code with CI, did work with only Azure and GCP provider, the test did fail for AWS provider. The logs for **successful test can be [viewed here]**(https://github.com/parul5sahoo/test/runs/7630080703?check_suite_focus=true) and the **logs for failed test can be [viewed here]**(https://github.com/parul5sahoo/test/runs/7626021380?check_suite_focus=true)

Signed-off-by: parul5sahoo [parulsahoo5jan@gmail.com](mailto:parulsahoo5jan@gmail.com)

### CODE REVIEW

File: test/framework/provider/wait.go [hunk 1]
```diff
@@ -27,20 +28,32 @@ import (
...
 
 // Wait for Provider to be successfully installed.
-func WaitForAllProvidersInstalled(ctx context.Context, c client.Client, interval time.Duration, timeout time.Duration) error {
+func WaitForAllProvidersInstalled(ctx context.Context, c client.Client, interval time.Duration, timeout time.Duration, t *testing.T) error {
 	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
 		l := &v1.ProviderList{}
 		if err := c.List(ctx, l); err != nil {
 			return false, err
 		}
+		// t.Logf("Provider list: %+v", l.Items)
```
Comment (hasheddan) [inline test/framework/provider/wait.go]:
```suggestion
```

File: test/framework/provider/wait.go [hunk 2]
```diff
@@ -52,27 +65,41 @@ func WaitForAllProvidersInstalled(ctx context.Context, c client.Client, interval
...
 
 // Wait for Provider to be successfully updated.
-func WaitForRevisionTransition(ctx context.Context, c client.Client, p2 string, p1 string, interval time.Duration, timeout time.Duration) error {
+func WaitForRevisionTransition(ctx context.Context, c client.Client, p2 string, p1 string, interval time.Duration, timeout time.Duration, t *testing.T) error {
 	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
 		l := &v1.ProviderRevisionList{}
 		if err := c.List(ctx, l); err != nil {
 			return false, err
 		}
+		// t.Logf("Provider Revisions list: %+v", l.Items)
```
Comment (hasheddan) [inline test/framework/provider/wait.go]:
```suggestion
```

File: test/framework/provider/wait.go [hunk 3]
```diff
@@ -52,27 +65,41 @@ func WaitForAllProvidersInstalled(ctx context.Context, c client.Client, interval
...
+		for _, item := range l.Items {
+			t.Logf("Name: %v \t Type: %v", item.Name, item.TypeMeta)
+		}
+
 		// There should be a revision present for the initial revision and the upgrade.
 		if len(l.Items) != 2 {
+			t.Log("The no. of providers revisions is not equal to 2")
```
Comment (hasheddan) [inline test/framework/provider/wait.go]:
```suggestion
			t.Log("The no. of provider revisions is not equal to 2")
```

File: test/framework/provider/wait.go [hunk 4]
```diff
@@ -84,12 +111,18 @@ func WaitForRevisionTransition(ctx context.Context, c client.Client, p2 string,
...
 
 // Wait for Provider to be successfully deleted.
-func WaitForAllProvidersDeleted(ctx context.Context, c client.Client, interval time.Duration, timeout time.Duration) error {
+func WaitForAllProvidersDeleted(ctx context.Context, c client.Client, interval time.Duration, timeout time.Duration, t *testing.T) error {
 	return wait.PollImmediate(interval, timeout, func() (bool, error) {
 		l := &v1.ProviderList{}
 		if err := c.List(ctx, l); err != nil {
 			return false, err
 		}
+		// t.Logf("Remaining Providers: %+v", l.Items)
```
Comment (hasheddan) [inline test/framework/provider/wait.go]:
```suggestion
```

File: test/framework/provider/wait.go [hunk 5]
```diff
@@ -52,27 +65,41 @@ func WaitForAllProvidersInstalled(ctx context.Context, c client.Client, interval
...
+func WaitForRevisionTransition(ctx context.Context, c client.Client, p2 string, p1 string, interval time.Duration, timeout time.Duration, t *testing.T) error {
 	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
 		l := &v1.ProviderRevisionList{}
 		if err := c.List(ctx, l); err != nil {
 			return false, err
 		}
+		// t.Logf("Provider Revisions list: %+v", l.Items)
+		for _, item := range l.Items {
+			t.Logf("Name: %v \t Type: %v", item.Name, item.TypeMeta)
+		}
```
Comment (hasheddan) [inline test/framework/provider/wait.go]: @parul5sahoo given that we are looping through items in cases below where we return `false` (i.e. we will keep polling), we could avoid doing so here as the later logs will inform us of the status that is causing us to continue polling.

File: test/framework/provider/wait.go [hunk 6]
```diff
@@ -52,27 +65,41 @@ func WaitForAllProvidersInstalled(ctx context.Context, c client.Client, interval
...
 			return false, nil
 		}
-		for _, p := range l.Items {
+		for _, item := range l.Items {
 			// New ProviderRevision should be Active.
-			if p.Spec.Package == p2 && p.GetDesiredState() != v1.PackageRevisionActive {
+			if item.Spec.Package == p2 && item.GetDesiredState() != v1.PackageRevisionActive {
+				t.Logf("The state of new provider revision %v having version %v is %v", item.Name, item.ObjectMeta.ResourceVersion, item.Status)
+				//TODO
+				// add on which p are we looking at using t.Logf.. for all statements
```
Comment (hasheddan) [inline test/framework/provider/wait.go]: Has this TODO been addressed?
```suggestion
```

**Review by hasheddan — COMMENTED** [review]
Thanks for your work here @parul5sahoo! Left some comments below, let me know if you have any questions :)

File: test/framework/provider/wait.go (hunk 6 above)
Comment (parul5sahoo) [reply test/framework/provider/wait.go]: Yes, this todo is addressed as the value we receive from `item.Name` gives us which provider revision we are working with.  Just forgot to remove the additional comments that I had kept in the file for my own reference. Removed them.

File: test/framework/provider/wait.go (hunk 5 above)
Comment (parul5sahoo) [reply test/framework/provider/wait.go]: Got it! Done!

Comment (parul5sahoo) [reply test/framework/provider/wait.go]: @hasheddan Should I do the same for [these lines](https://github.com/parul5sahoo/test/blob/6eb3d299a5de602da3e9f1b8dd452bbb25395419/test/framework/provider/wait.go#L37-L39) as well?

**Review by hasheddan — APPROVED** [review]
Thanks @parul5sahoo! Would you mind squashing the merge commit to clean up the history before merge here?

**Comment by parul5sahoo** [comment]
> Thanks @parul5sahoo! Would you mind squashing the merge commit to clean up the history before merge here?

Squashed the commits.

File: test/framework/provider/wait.go [hunk 7]
```diff
@@ -52,27 +61,35 @@ func WaitForAllProvidersInstalled(ctx context.Context, c client.Client, interval
...
+				t.Logf("Provider revision %v : %v uses package %v", i, item.Name, item.Spec.Package)
+			}
 			return false, nil
 		}
-		for _, p := range l.Items {
+		for _, item := range l.Items {
 			// New ProviderRevision should be Active.
-			if p.Spec.Package == p2 && p.GetDesiredState() != v1.PackageRevisionActive {
+			if item.Spec.Package == p2 && item.GetDesiredState() != v1.PackageRevisionActive {
+				t.Logf("The state of new provider revision %v having version %v is %v", item.Name, item.ObjectMeta.ResourceVersion, item.Status)
```
Comment (jbw976) [inline test/framework/provider/wait.go]: this `item.ObjectMeta.ResourceVersion` field is the Kubernetes API object version - it is not related to the Package version, so it isn't helpful in the logging.  See this logging statement from the [latest test run on this PR](https://github.com/crossplane/test/runs/7841241188?check_suite_focus=true):

```
    wait.go:92: The condition of provider revision provider-aws-b53fc91447be : 2791 is {[{Healthy False 2022-08-15 15:54:51 +0000 UTC UnhealthyPackageRevision }]}
```

The `ResourceVersion` here is 2791 and isn't super helpful :)

I think `item.Spec.Package` and `item.Spec.Revision` would be more useful to log here.  This is also applicable on all the other lines that we are using `item.ObjectMeta.ResourceVersion`, so make sure to update those too.

File: test/framework/provider/wait.go [hunk 8]
```diff
@@ -84,12 +101,17 @@ func WaitForRevisionTransition(ctx context.Context, c client.Client, p2 string,
...
-func WaitForAllProvidersDeleted(ctx context.Context, c client.Client, interval time.Duration, timeout time.Duration) error {
+func WaitForAllProvidersDeleted(ctx context.Context, c client.Client, interval time.Duration, timeout time.Duration, t *testing.T) error {
 	return wait.PollImmediate(interval, timeout, func() (bool, error) {
 		l := &v1.ProviderList{}
 		if err := c.List(ctx, l); err != nil {
 			return false, err
 		}
+		for _, item := range l.Items {
+			t.Log("Undeleted providers :")
+			t.Logf("Name: %v \t Type: %v", item.Name, item.Kind)
```
Comment (jbw976) [inline test/framework/provider/wait.go]: `item.Spec.Package` would be useful to log here also

**Review by jbw976 — COMMENTED** [review]
Thanks for the discussion today on this PR @parul5sahoo and @hasheddan! 

I have a couple suggestions to improve the logging details, see if you can incorporate those and we'll get this merged.

It's nice that the provider upgrade tests are failing consistently for provider-aws, we get to see your logging changes in action haha

File: test/framework/provider/wait.go (hunk 8 above)
Comment (parul5sahoo) [reply test/framework/provider/wait.go:112]: Do you mean logging `item.Spec.Package` in addition to the `item.Name` and `item.Kind`? @jbw976

File: test/framework/provider/wait.go (hunk 7 above)
Comment (parul5sahoo) [reply test/framework/provider/wait.go]: Ohkayy! got it

File: .github/workflows/crossplane-upgrade.yml [hunk 9]
```diff
@@ -33,7 +33,7 @@ jobs:
         run: kubectl create namespace crossplane-system
         shell: bash
       - name: Install Crossplane from Stable
-        run: helm repo add crossplane-stable https://charts.crossplane.io/stable && helm repo update && helm install crossplane --namespace crossplane-system crossplane-stable/crossplane --version 1.1.0 --wait
+        run: helm repo add crossplane-stable https://charts.crossplane.io/stable && helm repo update && helm install crossplane --namespace crossplane-system crossplane-stable/crossplane --version 1.9.0 --wait
```
Comment (jbw976) [inline .github/workflows/crossplane-upgrade.yml]: any idea why this change is showing up again? I thought we had already merged this into master as part of #24 🤔

File: test/framework/provider/wait.go [hunk 10]
```diff
@@ -52,27 +61,35 @@ func WaitForAllProvidersInstalled(ctx context.Context, c client.Client, interval
...
 			// Old ProviderRevision should be Inactive.
-			if p.Spec.Package == p1 && p.GetDesiredState() != v1.PackageRevisionInactive {
+			if item.Spec.Package == p1 && item.GetDesiredState() != v1.PackageRevisionInactive {
+				t.Logf("The state of old provider revision %v built from package %v and having version %v is %v", item.Name, item.Spec.Package, item.Spec.Revision, item.Status)
 				return false, nil
 			}
 			// Both ProviderRevisions should be healthy.
-			if p.GetCondition(v1.TypeHealthy).Status != corev1.ConditionTrue {
+			if item.GetCondition(v1.TypeHealthy).Status != corev1.ConditionTrue {
+				t.Logf("The condition of provider revision %v: %v built from package %v  is %v", item.Name, item.Spec.Revision, item.Spec.Package, item.Status.ConditionedStatus)
```
Comment (jbw976) [inline test/framework/provider/wait.go]: can you make the format of this logging statement match the other 2? the arguments are reversed in this one, but keeping them consistent would be easier to read/parse in the logs if the test goes through more than 1 of these temporary failure/waiting conditions

```suggestion
				t.Logf("The condition of provider revision %v built from package %v and having version %v is %v", item.Name, item.Spec.Package, item.Spec.Revision, item.Status.ConditionedStatus)
```

**Review by jbw976 — COMMENTED** [review]
1 more logging clean up to do, and I'm definitely curious about the changes showing in this PR's diff that should already be in master - not sure what to make of that 😨

**Comment by jbw976** [comment]
So cool to see how the logging looks after each iteration, and try to actually debug a failing CI test run: https://github.com/crossplane/test/runs/7880091512?check_suite_focus=true 💪

File: test/framework/provider/wait.go (hunk 10 above)
Comment (parul5sahoo) [reply test/framework/provider/wait.go]: Done!

File: .github/workflows/crossplane-upgrade.yml (hunk 9 above)
Comment (parul5sahoo) [reply .github/workflows/crossplane-upgrade.yml]: That's probably because I did not pull the latest commits that were merged to master branch before creating the PR and it show up as another commit when the master branch was merged into logging branch. However, I was not able to squash that commit into the previous commit.

**Comment by parul5sahoo** [comment]
> So cool to see how the logging looks after each iteration, and try to actually debug a failing CI test run: https://github.com/crossplane/test/runs/7880091512?check_suite_focus=true muscle

It's again the failure due to AWS provider, because the [condition wasn't healthy](https://github.com/crossplane/test/runs/7880091512?check_suite_focus=true#step:9:122) and there was [a time out](https://github.com/crossplane/test/runs/7880091512?check_suite_focus=true#step:9:124)

**Review by jbw976 — APPROVED** [review]
awesome @parul5sahoo thanks for incorporating the final feedback! :)
